akka23-common/src/main/scala/com/typesafe/conductr/akka/ConnectionHandler.scala

```diff
@@ -70,39 +76,89 @@ class ConnectionHandler extends AbstractConnectionHandler {
 
   override protected type CC = ConnectionContext
 
-  override def withConnectedRequest[T](
-    payload: Option[HttpPayload])(thunk: (Int, Map[String, Option[String]]) => Option[T])(implicit cc: CC): Future[Option[T]] = {
-
+  override def withConnectedRequest[T](payload: Option[HttpPayload])(handler: (Int, Map[String, Option[String]]) => Option[T])(implicit cc: CC): Future[Option[T]] = {
     payload.fold[Future[Option[T]]](Future.successful(None)) { p =>
-      val connection = cc.httpExt.outgoingConnection(p.getUrl.getHost, p.getUrl.getPort)
-      val url = p.getUrl
-      val urlStr = url.toString
-
-      val requestMethod = p.getRequestMethod match {
-        case "GET"     => Get(urlStr)
-        case "POST"    => Post(urlStr)
-        case "PUT"     => Put(urlStr)
-        case "PATCH"   => Patch(urlStr)
-        case "DELETE"  => Delete(urlStr)
-        case "OPTIONS" => Options(urlStr)
-        case "HEAD"    => Head(urlStr)
-      }
+      val connection = createConnection(p)
+      val request = createRequest(p)
 
-      val request = requestMethod
-        .addHeader(`User-Agent`(UserAgent))
-        .addHeader(Host(url.getHost))
-
-      val requestSource = Source.single(request)
+      Source.fromFuture(request)
         .via(connection)
         .map { response =>
-          thunk(
+          handler(
             response.status.intValue(),
             response.headers.foldLeft(Map.empty[String, Option[String]]) {
               case (m, header) => m.updated(header.name(), Some(header.value()))
             })
         }
-      requestSource.runWith(Sink.head)(cc.actorMaterializer)
+        .runWith(Sink.head)(cc.actorMaterializer)
+    }
+  }
+
+  def withConnectedRequest[T](payload: HttpPayload, body: Option[Future[RequestEntity]] = None)(handler: (Int, Map[String, Option[String]], ResponseEntity) => Future[T])(implicit cc: CC): Future[T] = {
```

Comment:
Would it be hard to do this now i.e. have `HttpPayload` take  a body param which is a `org.reactivestreams.Publisher<T>`? I'm completely fine having reactivestreams as a dependency for Java given that it will become a part of Java.